crates/linker/src/pe.rs

```diff
@@ -1406,9 +1258,11 @@ fn relocate_dummy_dll_entries(executable: &mut [u8], md: &PeMetadata) {
                 * std::mem::size_of::<pe::ImageDataDirectory>(),
     );
 
-    // it is crucial that the directory size is rounded up to a multiple of 8!
-    // the value is already rounded, so to be correct we can't rely on `dir.size.get(LE)`
-    let new_reloc_directory_size = next_multiple_of(base_relocations.new_size as usize, 8);
+    // for unclear reasons, we must bump the image directory size here.
+    // we also need some zeroed-out memory at the end, so if the directory
+    // ends at a multiple of `file_alignment`, pick the next one.
+    let new_reloc_directory_size =
+        next_multiple_of(dir.size.get(LE) as usize + 4, md.file_alignment as usize);
     dir.size.set(LE, new_reloc_directory_size as u32);
```

Comment:
it looks like for the relocation image to work correctly,there need to be at least 4 zeros after the last entry (kind of like a null sentinel). But if the current relocations already take up all available space, then we must add some extra null bytes before starting the next section.

It can (and did) happen that the `.reloc` section was completely used (no free space for the 0's), but it was the final section of the host binary, so not expanded beyond that size, I guess EOF is also a valid sentinel. But when we add more sections, the bytes after the `.reloc` section are non-zero, and we run into trouble.